peps/pep-0797.rst

```diff
@@ -0,0 +1,672 @@
+
+The only place where this needs to change is in
+:c:func:`PyUnicode_CHECK_INTERNED`. If the current state is
+``SSTATE_INTERNED_MORTAL``, but the object is immortal, then it should
+return ``SSTATE_INTERNED_IMMORTAL`` instead.
+
+Changes to Object Finalization
+******************************
+
+The C API function, :c:func:`PyObject_CallFinalizer`, must be able to
+de-duplicate calls to immortal objects, even if they're non-GC types.
+
+Immortalization APIs
+********************
+
+This PEP introduces two new semi-public APIs for making an object immortal:
+:c:func:`PyUnstable_Immortalize` and :func:`sys._immortalize`.
+
+.. c:function:: int PyUnstable_Immortalize(PyObject *obj)
+
+   Make *obj* :term:`immortal`. An :term:`immortal` object can have higher performance for
+   :term:`reference count` modifications with the cost of being less memory efficient.
+
+   This function returns ``1`` if *obj* was successfully made immortal, or
+   ``0`` if the object is already immortal. On error, this function returns
+   ``-1`` with an exception set. The caller must hold an
+   :term:`attached thread state`.
+
+   In order to be made immortal, an object must follow the "immortalization
+   contract". This contract is based on best practices in Python's C API, and
+   it's very unlikely that a type doesn't follow it:
+
+   * The object must be allocated under either the "object" or "memory" allocator
+     domains. See :ref:`Allocator Domains <python:allocator-domains>` for more information.
+     Default values of :c:member:`~PyTypeObject.tp_alloc` always use the object
+     allocator.
+   * All :term:`strong references <strong reference>` released in the object's
+     :c:member:`~PyTypeObject.tp_dealloc` must also be traversed by the
+     object's :c:member:`~PyTypeObject.tp_traverse` slot, if the type can
+     contain circular references. This is a general requirement for any garbage
+     collected type.
+   * All finalization for an object must be done in
+     :c:member:`~PyTypeObject.tp_finalize` by
+     :c:func:`PyObject_CallFinalizerFromDealloc`.
+
+   .. warning::
+      Abuse of this function can result in high memory usage for a program.
+
+.. function:: sys._immortalize(obj)
+
+   Make *obj* an :term:`immortal` object. This means that it will not be deallocated for the
+   lifetime of the Python interpreter.
+
+   For objects that have costly modifications to their :term:`reference count`,
+   making them :term:`immortal` can help improve performance. For example, on
+   the :term:`free-threaded <free threading>` build, making an object
+   :term:`immortal` can help mitigate bottlenecks that come from concurrent
+   modification of an object's :term:`reference count`.
+
+   .. warning::
+      This function should be considered a low-level routine that most users should avoid. Making too many
+      objects immortal can result in drastically higher memory usage for a program.
+
+Immortalization Contract
+************************
```

Comment:
This topic & idea is introduced in *Specification*. Can you move the bits that already exist today to an earlier section, so that the Specification only needs to detail the changes or strengthening to the contract?